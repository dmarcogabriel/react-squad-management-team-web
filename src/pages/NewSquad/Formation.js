import React, { useState, useEffect } from 'react';
import Player from './Player';

const Formation = ({ formation = [], onDropPlayer = () => {} }) => {
  const [rows, setRows] = useState([]);

  useEffect(() => {
    setRows(formation);
  }, [formation]);

  return (
    <div className="bg-gradient-to-t from-secondary-dark to-secondary-light rounded-sm p-2 mb-5 justify-center">
      {rows.map(({ players, rowNumber, id }, i) => (
        <div
          className={`grid grid-cols-3 grid-rows-${rowNumber > 3 ? '2' : '1'}`}
          key={`r${String(i)}`}
        >
          {players.map((player) => (
            <div key={player.position} className="m-2 text-center">
              <Player
                player={player}
                onDropPlayer={(playerId) =>
                  onDropPlayer(playerId, id, player.position)
                }
              />
            </div>
          ))}
        </div>
      ))}
    </div>
  );
};

export default Formation;
