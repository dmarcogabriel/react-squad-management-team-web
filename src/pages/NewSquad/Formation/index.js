import React from 'react';
import cn from 'classnames';
import Player from '../Player';

const Formation = ({ formation = [], onDropPlayer }) => {
  return (
    <div
      data-testid="formation"
      className={cn(
        'bg-gradient-to-t from-secondary-dark to-secondary-light',
        'rounded-sm p-2 mb-5 justify-center'
      )}
    >
      {formation.map(({ players, rowNumber, id }, i) => (
        <div
          data-testid={`formation-row-${id}`}
          className={`grid grid-cols-3 grid-rows-${rowNumber > 3 ? '2' : '1'}`}
          key={`r${String(i)}`}
        >
          {players.map((player) => (
            <div
              data-testid={`formation-player-${player.position}`}
              key={player.position}
              className="m-2 text-center"
            >
              <Player
                player={player}
                onDropPlayer={(playerId) =>
                  onDropPlayer(playerId, id, player.position)
                }
              />
            </div>
          ))}
        </div>
      ))}
    </div>
  );
};

export default Formation;
