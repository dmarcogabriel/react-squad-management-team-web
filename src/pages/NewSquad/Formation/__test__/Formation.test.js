import ReactDOM from 'react-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import Formation from '..';

const Component = (props) => (
  <DndProvider backend={HTML5Backend}>
    <Formation {...props} />
  </DndProvider>
);

const mockFormation352 = [
  {
    id: 'c1',
    colNumber: 3,
    players: [
      {
        position: 'pl1',
        player: null,
      },
      {
        position: 'pl2',
        player: null,
      },
      {
        position: 'pl3',
        player: null,
      },
    ],
  },
  {
    id: 'c2',
    colNumber: 2,
    players: [
      {
        position: 'pl1',
        player: null,
      },
      {
        position: 'pl12',
        player: null,
      },
      {
        position: 'pl3',
        player: null,
      },
      {
        position: 'pl4',
        player: null,
      },
      {
        position: 'pl5',
        player: null,
      },
    ],
  },
  {
    id: 'c3',
    colNumber: 2,
    players: [
      {
        position: 'pl1',
        player: null,
      },
      {
        position: 'pl12',
        player: null,
      },
    ],
  },
];

const mockFormation3223 = [
  {
    id: 'c1',
    colNumber: 3,
    players: [
      {
        position: 'pl1',
        player: null,
      },
      {
        position: 'pl2',
        player: null,
      },
      {
        position: 'pl3',
        player: null,
      },
    ],
  },
  {
    id: 'c2',
    colNumber: 2,
    players: [
      {
        position: 'pl4',
        player: null,
      },
      {
        position: 'pl5',
        player: null,
      },
    ],
  },
  {
    id: 'c3',
    colNumber: 2,
    players: [
      {
        position: 'pl6',
        player: null,
      },
      {
        position: 'pl7',
        player: null,
      },
    ],
  },
  {
    id: 'c4',
    colNumber: 2,
    players: [
      {
        position: 'pl8',
        player: null,
      },
      {
        position: 'pl9',
        player: null,
      },
      {
        position: 'pl10',
        player: null,
      },
    ],
  },
];

describe('<Formation />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Component />, div);
  });

  it('renders formation 3-2-2-3 correctly', () => {
    const { getByTestId } = render(<Component formation={mockFormation3223} />);

    const row = getByTestId('formation');
    const row1 = getByTestId('formation-row-c1');
    const row2 = getByTestId('formation-row-c2');
    const row3 = getByTestId('formation-row-c3');
    const row4 = getByTestId('formation-row-c4');

    expect(row.childElementCount).toEqual(4);

    expect(row1.childElementCount).toEqual(3);
    expect(row2.childElementCount).toEqual(2);
    expect(row3.childElementCount).toEqual(2);
    expect(row4.childElementCount).toEqual(3);
  });

  it('renders formation 3-5-2 correctly', () => {
    const { getByTestId } = render(<Component formation={mockFormation352} />);

    const row = getByTestId('formation');
    const row1 = getByTestId('formation-row-c1');
    const row2 = getByTestId('formation-row-c2');
    const row3 = getByTestId('formation-row-c3');

    expect(row.childElementCount).toEqual(3);

    expect(row1.childElementCount).toEqual(3);
    expect(row2.childElementCount).toEqual(5);
    expect(row3.childElementCount).toEqual(2);
  });

  // it('executes drop event correctly', () => {
  //   const mockCallback = jest.fn((playerId, id, playerPosition) => ({
  //     playerId,
  //     id,
  //     playerPosition,
  //   }));
  //   const mockPlayer = {
  //     player_id: 'm1',
  //     fullname: 'Mock Player',
  //     age: 24,
  //     nationality: 'Mockland',
  //   };

  //   const { getByTestId } = render(
  //     <Component formation={mockFormation3223} onDropPlayer={mockCallback} />
  //   );

  //   const player = getByTestId('formation-player-pl1');

  //   fireEvent.drop(player, {
  //     dataTransfer: {
  //       files: [<DraggablePlayer player={mockPlayer} />, { type: 'player' }],
  //     },
  //   });
  // });

  it('matches snapshot', () => {
    expect(
      renderer.create(<Component formation={mockFormation352} />).toJSON()
    ).toMatchSnapshot();
  });
});
