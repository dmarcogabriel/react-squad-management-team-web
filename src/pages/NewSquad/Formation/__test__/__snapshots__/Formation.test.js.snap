// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Formation /> matches snapshot 1`] = `
<div
  className="bg-gradient-to-t from-secondary-dark to-secondary-light rounded-sm p-2 mb-5 justify-center"
/>
`;
